### PR TITLE
[CI] Use windows docker image in CI.

### PR DESCRIPTION
**Pull request by larshg**

### CODE REVIEW

**Comment by kunaltyagi** [comment]
Perhaps, add a dockerfile and a CI to build the image so you can use PCL's dockerhub to maintain the images

**Comment by larshg** [comment]
Yeah, I'll do that when it works. Currently it fail to access `c:\__a\bin` and I have no clue why it goes there. And can't find anyone else having the same problem...

**Comment by larshg** [comment]
So how do I go about updating the build YAML - should I add a new job as the windows container has to be build on a windows vmImage?

**Comment by kunaltyagi** [comment]
> So how do I go about updating the build YAML 

Modify the .ci/azure-pipelines/env.yaml to add a new job (with windows as the pool's vmImage) to build the windows docker image.

File: cmake/pcl_options.cmake [hunk 1]
```diff
@@ -83,3 +83,6 @@ endif()
 # (Used to prevent gpu tests from executing in CI where GPU hardware is unavailable)
 option(PCL_DISABLE_GPU_TESTS "Disable running GPU tests. If disabled, tests will still be built." OFF)
 
+# Set whether visualizations tests should be run
+# (Used to prevent visualizations tests from executing in CI where visualization is unavailable)
+option(PCL_DISABLE_GPU_TESTS "Disable running visualizations tests. If disabled, tests will still be built." OFF)
```
Comment (kunaltyagi) [inline cmake/pcl_options.cmake:88]: Error in naming?

File: .ci/azure-pipelines/azure-pipelines.yaml [hunk 2]
```diff
@@ -126,25 +130,28 @@ stages:
...
-          vmImage: 'vs2017-win2016'
+          vmImage: 'windows-2019'
         strategy:
           matrix:
             x86:
+              CONTAINER: winx86
               PLATFORM: 'x86'
               ARCHITECTURE: 'x86'
-              GENERATOR: 'Visual Studio 15 2017'
+              GENERATOR: '"Visual Studio 16 2019" -A Win32'
```
Comment (kunaltyagi) [inline .ci/azure-pipelines/azure-pipelines.yaml:143]: Why are we moving from Visual Studio 15? Is it EOL?

Comment (larshg) [reply .ci/azure-pipelines/azure-pipelines.yaml:143]: No, I'm just used to use VS16, but maybe we should have the x86 run VS15 and x64 run VS16?

Comment (kunaltyagi) [reply .ci/azure-pipelines/azure-pipelines.yaml:143]: Either is fine, TBH. It's just x86 is consistently failing since a few PRs (even this)

Comment (larshg) [reply .ci/azure-pipelines/azure-pipelines.yaml:143]: Can see it makes a segfault. We would have to look into  that.

Comment (larshg) [reply .ci/azure-pipelines/azure-pipelines.yaml:157]: This error is fixed now.

**Review by kunaltyagi — APPROVED** [review]
Concerns due to move from Visual Studio 2017 to 2019:
* Is it because of the use of windows 2019 container OS?
* Is the 2019 version the "default" one used now? (I don't know how versions are maintained for VS)
* If we are dropping support for the older versions, we might want to check if we have some hacks for vs2017 or earlier

Rest LGTM.

We can finally compile visualization on Windows (in same time as before) 🎉 🥳

**Comment by larshg** [comment]
> Concerns due to move from Visual Studio 2017 to 2019:
> 
> * Is it because of the use of windows 2019 container OS?
> * Is the 2019 version the "default" one used now? (I don't know how versions are maintained for VS)
> * If we are dropping support for the older versions, we might want to check if we have some hacks for vs2017 or earlier
> 
> Rest LGTM.
> 
> We can finally compile visualization on Windows (in same time as before) 🎉 🥳

* No, it was probably due to the tutorials I found, used the newer version, but we can probably install VS2017 if we want to.
* If you search for Visual Studio, you will be directed to a 2019 download. There is nothing default preinstalled.
* Right now we don't "support" VS2019, so either one is unsupported, unless we install one of them with VS2017 and the other with VS2019.

VS2019 seems also to be overrepresented in many of the relatively new opened windows issues.

File: test/CMakeLists.txt [hunk 3]
```diff
@@ -23,11 +23,18 @@ if(MSVC)
...
+  string(APPEND EXCLUDE_TESTS visualization)
+endif()
+
 # Enables you to disable GPU tests. Used on CI as it has no access to GPU hardware
 if(PCL_DISABLE_GPU_TESTS)
-  set(PCL_CTEST_ARGUMENTS ${PCL_CTEST_ARGUMENTS} -E gpu)
+  string(APPEND EXCLUDE_TESTS gpu)
 endif()
 
+set(PCL_CTEST_ARGUMENTS ${PCL_CTEST_ARGUMENTS} -E ${EXCLUDE_TESTS})
```
Comment (mvieth) [inline test/CMakeLists.txt:36]: `-E` is added even if no test is disabled. Also some quotation marks might be needed here to ensure correct whitespace. Do you know whether it has to be `-E gpu -E visualization` or `-E gpu visualization`? A quick internet search did not give any answers

Comment (kunaltyagi) [reply test/CMakeLists.txt:36]: Good point. `-E regex` seems to be the API, so I'd guess it means `-E gpu` is correct, not `-E gpu viz`. I don't know if we can have 2 -E though

Comment (larshg) [reply test/CMakeLists.txt:36]: I just know that before this [commit](https://github.com/PointCloudLibrary/pcl/pull/4426/commits/4e97364101c466e0814548309837e41e39c8110c), only the GPU was exluded and it failed running the visualization test. After the commit it correctly skipped both.
I can have a look at it again. I guess if its a regex, it should be possible to match one or the other, so they are correctly skipped.

Comment (kunaltyagi) [reply test/CMakeLists.txt:36]: I think the bad documentation of CTest aids in exploration and exploitation of any behavior, as per Hyrum's law 

😆 If it works, I guess it's fine....

Comment (larshg) [reply test/CMakeLists.txt]: I have updated to properly build the regex.

**Review by kunaltyagi — APPROVED** [review]
Looks quite wonderful

**Review by mvieth — APPROVED** [review]

**Comment by mvieth** [comment]
@larshg Merge?
Maybe the title could be changed

**Comment by larshg** [comment]
Sound more reasonable?

File: test/CMakeLists.txt [hunk 4]
```diff
@@ -23,9 +23,21 @@ if(MSVC)
...
 if(PCL_DISABLE_GPU_TESTS)
-  set(PCL_CTEST_ARGUMENTS ${PCL_CTEST_ARGUMENTS} -E gpu)
+  list(APPEND EXCLUDE_TESTS gpu)
+endif()
+
+#Check if there are any tests to exclude
+if(EXCLUDE_TESTS)
+  message(STATUS "Tests excluded: ${EXCLUDE_TESTS}")
+  string(REPLACE ";" "|" EXCLUDE_TESTS_REGEX "${EXCLUDE_TESTS}")
+  set(PCL_CTEST_ARGUMENTS ${PCL_CTEST_ARGUMENTS} -E "(${EXCLUDE_TESTS})")
```
Comment (larshg) [inline test/CMakeLists.txt:40]:
```suggestion
  set(PCL_CTEST_ARGUMENTS ${PCL_CTEST_ARGUMENTS} -E "(${EXCLUDE_TESTS_REGEX})")
```

Comment (larshg) [reply test/CMakeLists.txt:40]: This seems like a typo, but the test was still excluded 😕

**Review by kunaltyagi — APPROVED** [review]

**Review by mvieth — APPROVED** [review]

**Comment by larshg** [comment]
Rebased and squashed a few commits. So should be fine to 🚀 when CI is green.

**Review by kunaltyagi — APPROVED** [review]

**Review by mvieth — APPROVED** [review]

**Pull request merged**